dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/Maturity.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2004-2024, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.common;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Can be used to mark endpoint methods with their maturity level. If types are annotated all
+ * endpoints of the type inherit the level unless their method has an overriding annotation.
+ *
+ * <p>Can also be used to mark endpoint parameters with their maturity level. In parameter objects
+ * the corresponding fields are annotated.
+ *
+ * @author Jan Bernitt
+ */
+@Target({ElementType.METHOD, ElementType.TYPE, ElementType.PARAMETER, ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Maturity {
+
+  enum Degree {
+    /**
+     * The API is mature, 3rd parties can depend on it and expect that breaking changes will occur
+     * rarely and only after a deprecation period.
+     */
+    STABLE,
+    /**
+     * The API has not matured yet. It might be changed or removed later on. 3rd parties should not
+     * rely on it and use this at their own risk.
+     */
+    EXPERIMENTAL,
+    /**
+     * 3rd parties should never depend on this, it is for internal use only. This means apps
+     * supported by the DHIS project use this, but it is subject to change whenever and however the
+     * app see fit.
+     */
+    INTERNAL
```

Comment:
I'm just wondering if INTERNAL should be part of this enum.
Maybe I'm overthinking, but we could also have something like:

`enum Exposure {
  INTERNAL,
  PUBLIC
}`

For example:
an API can be INTERNAL, and EXPERIMENTAL (or STABLE). It could also be PUBLIC and BETA.

Also, I'm unsure if INTERNAL is a state of maturity itself.

Reply:
I'd agree with Maikel with regard to `INTERNAL` not being a maturity state. It's more about who should use the API.

Reply:
Assuming we keep things in 1 dimension of classification this is mostly a problem of the naming. If we name the enum and constant differently we might find names that are (more) coherent. That would be my goal.

Reply:
What if we call this enum `Category`, for example? This is more generic and would capture any of those intentions.

Reply:
It seems to me that it is indeed just a naming issue right now.
With 3 different statuses we should be covered and their meaning are:
- Safe to use
- Try it but do not get too comfortable (it could become safe to use )
- Not safe to use (it will never become safe to use )

I would suggest instead of `INTERNAL` something like `UNSTABLE` but I am not sure because it gives the idea that could change to `STABLE` at some point.

Reply:
Yes, I think @enricocolasante found a good summary and way to put it. So we need better naming to reflect this and ideally make that read coherent. 

`Category` is a good idea, except the name clash with our metadata category makes it a bit confusing. Might still be better.

Reply:
> We could consider assigning a score instead where 1 is the least mature and 3 is the most mature.

Good thinking outside the box I accidentally created, but a score would give the idea of linearity or a degree to which something is given. This does not match well with the 3 classifications we proposed thus far which are different and not by any means on some axis except that they are mutually exclusive.

Reply:
I agree we should not use "INTERNAL", as one of our core ideas with the API is that it should all be public and consumable by others, and any app, script or interaction could be recreated by external people.

Switching it over to unstable sounds a lot better, but it's important that we use any non-stable maturity only in special cases, so we don't end up marking everything as unstable or experimental.

I would argue that experimental APIs are APIs where we dont have the complete implementation done (IE. we only implemented parts of the API, since we had to add partial support for something). Likewise, unstable APIs would be APIs that are new, and might not have any adoption in our core apps, and are subject to be changed before transitioning to stable.